pkg/provider/azure_local_services.go

```diff
@@ -218,3 +275,272 @@ func newBatchOperationResult(name string, success bool, err error) batchOperatio
+func (az *Cloud) getBackendPoolNamesForService(service *v1.Service, clusterName string) map[bool]string {
+	if !isLocalService(service) || !az.useMultipleStandardLoadBalancers() {
+		return getBackendPoolNames(clusterName)
+	}
+	return map[bool]string{
+		consts.IPVersionIPv4: getLocalServiceBackendPoolName(getServiceName(service), false),
+		consts.IPVersionIPv6: getLocalServiceBackendPoolName(getServiceName(service), true),
+	}
+}
+
+// getBackendPoolIDsForService determine the expected backend pool IDs
+// by checking the external traffic policy of the service.
+func (az *Cloud) getBackendPoolIDsForService(service *v1.Service, clusterName, lbName string) map[bool]string {
+	if !isLocalService(service) || !az.useMultipleStandardLoadBalancers() {
+		return az.getBackendPoolIDs(clusterName, lbName)
+	}
+	return map[bool]string{
+		consts.IPVersionIPv4: az.getLocalServiceBackendPoolID(getServiceName(service), lbName, false),
+		consts.IPVersionIPv6: az.getLocalServiceBackendPoolID(getServiceName(service), lbName, true),
+	}
+}
+
+// getLocalServiceBackendPoolID gets the ID of the backend pool of a local service.
+func (az *Cloud) getLocalServiceBackendPoolID(serviceName string, lbName string, ipv6 bool) string {
+	return az.getBackendPoolID(lbName, getLocalServiceBackendPoolName(serviceName, ipv6))
+}
+
+// localServiceOwnsBackendPool checks if a backend pool is owned by a local service.
+func localServiceOwnsBackendPool(serviceName, bpName string) bool {
+	prefix := strings.Replace(serviceName, "/", "-", -1)
+	return strings.HasPrefix(strings.ToLower(bpName), strings.ToLower(prefix))
+}
+
+type serviceInfo struct {
+	ipFamily string
+	lbName   string
+}
+
+func newServiceInfo(ipFamily, lbName string) *serviceInfo {
+	return &serviceInfo{
+		ipFamily: ipFamily,
+		lbName:   lbName,
+	}
+}
+
+// getLocalServiceEndpointsNodeNames gets the node names that host all endpoints of the local service.
+func (az *Cloud) getLocalServiceEndpointsNodeNames(service *v1.Service) (sets.Set[string], error) {
+	var ep *discovery_v1.EndpointSlice
+	az.endpointSlicesCache.Range(func(key, value interface{}) bool {
+		endpointSlice := value.(*discovery_v1.EndpointSlice)
+		if strings.EqualFold(getServiceNameOfEndpointSlice(endpointSlice), service.Name) &&
+			strings.EqualFold(endpointSlice.Namespace, service.Namespace) {
+			ep = endpointSlice
+			return false
+		}
+		return true
+	})
+	if ep == nil {
+		klog.Infof("EndpointSlice for service %s/%s not found, try to list EndpointSlices", service.Namespace, service.Name)
+		eps, err := az.KubeClient.DiscoveryV1().EndpointSlices(service.Namespace).List(context.Background(), metav1.ListOptions{})
+		if err != nil {
+			klog.Errorf("Failed to list EndpointSlices for service %s/%s: %s", service.Namespace, service.Name, err.Error())
+			return nil, err
+		}
+		for _, endpointSlice := range eps.Items {
+			endpointSlice := endpointSlice
+			if strings.EqualFold(getServiceNameOfEndpointSlice(&endpointSlice), service.Name) {
```

Comment:
Here we should also compare namespace.

Reply:
We use service name label on the endpoinslice to determine the ownership here.

Reply:
oh I see. We are listing es from the same ns.